### PR TITLE
add -verify option to checkfallbacks fix #3179

### LINKED ISSUE #3179: Regression test of ATS configuration
To avoid those edge cases like ERR_TOO_MANY_REDIRECTS from recurring. Should be able to test through single command, or automated. Maybe add it as a functionality of `chkfallbacks`.

### PR DESCRIPTION
**Pull request by fffw**
Fixes https://github.com/getlantern/lantern/issues/3179

This new option can be used to verify the functionality of ATS based chained server. Currently it only checks https://github.com/getlantern/lantern/issues/3133 and https://github.com/getlantern/lantern/issues/3147, but we can add more checks later.

### CODE REVIEW

**Comment by fffw** [comment]
@uaalto My PR comes out finally, mind take a look?

File: src/github.com/getlantern/checkfallbacks/checkfallbacks.go [hunk 1]
```diff
@@ -149,6 +150,7 @@ func testFallbackServer(fb *client.ChainedServerInfo, workerId int) (output full
 		output.info = []string{"\n" + string(reqStr)}
 	}
 
+	req.Header.Set("X-LANTERN-AUTH-TOKEN", fb.AuthToken)
```
Comment (uaalto) [inline src/github.com/getlantern/checkfallbacks/checkfallbacks.go]: Isn't this header added already in balancer? Otherwise checkfallbacks wouldn't be working ATM.

File: src/github.com/getlantern/checkfallbacks/verify.go [hunk 2]
```diff
@@ -0,0 +1,137 @@
...
+	}
+	req.Header.Set("X-LANTERN-AUTH-TOKEN", "invalid")
+	doVerifyMimic(fb.Addr, req, c)
+}
+
+func doVerifyMimic(addr string, req *http.Request, c *http.Client) {
+	resp, err := c.Do(req)
+	if err != nil {
+		log.Errorf("%v: requesting humans.txt failed: %v", addr, err)
+		if *verbose {
```
Comment (uaalto) [inline src/github.com/getlantern/checkfallbacks/verify.go]: So, with this we would also be able to print the request to the proxy, right? (Not just the request passed to it to relay, once it is connected). So, effectively we are bypassing the flashlight code currently being used to test fallbacks?

File: src/github.com/getlantern/checkfallbacks/verify.go [hunk 3]
```diff
@@ -0,0 +1,137 @@
...
+		log.Debugf("Body expected: %v", string(atsDefaultBody))
+		log.Debugf("Body got: %v", string(b))
+		return
+	}
+	log.Debugf("%v: OK.", addr)
+}
+
+func verifyRedirectSites(fb *client.ChainedServerInfo, c *http.Client, url string) {
+	req, err := http.NewRequest("GET", url, nil)
+	req.Header.Set("X-LANTERN-AUTH-TOKEN", fb.AuthToken)
```
Comment (uaalto) [inline src/github.com/getlantern/checkfallbacks/verify.go]: I think that if the previous line fails, this would segfault.

File: src/github.com/getlantern/checkfallbacks/verify.go [hunk 4]
```diff
@@ -0,0 +1,137 @@
...
+			log.Debug(string(reqStr))
+		}
+		return
+	}
+	defer func() {
+		if err := resp.Body.Close(); err != nil {
+			log.Debugf("Unable to close response body: %v", err)
+		}
+	}()
+	if resp.StatusCode != 200 {
```
Comment (uaalto) [inline src/github.com/getlantern/checkfallbacks/verify.go]: Is this the only code we might expect? Is it possible that a server returns 20x or 30x and still be acceptable?

File: src/github.com/getlantern/checkfallbacks/checkfallbacks.go (hunk 1 above)
Comment (fffw) [inline src/github.com/getlantern/checkfallbacks/checkfallbacks.go]: Interesting. During my test, `checkfallbacks` never succeeded without this line.

File: src/github.com/getlantern/checkfallbacks/verify.go (hunk 2 above)
Comment (fffw) [inline src/github.com/getlantern/checkfallbacks/verify.go]: Oh I didn't notice that I'm printing the original request. In fact it's somewhat meaningless to print request here if http connection failed.

File: src/github.com/getlantern/checkfallbacks/verify.go (hunk 3 above)
Comment (fffw) [inline src/github.com/getlantern/checkfallbacks/verify.go]: Good catch!

File: src/github.com/getlantern/checkfallbacks/checkfallbacks.go (hunk 1 above)
Comment (myleshorton) [inline src/github.com/getlantern/checkfallbacks/checkfallbacks.go]: balancer never touches anything with HTTP -- it just deals with net.Dial and net.Conn @fffw and @uaalto. That was actually part of the difficulty with adding the auth header more generally outside of using CONNECT for everything in chained/dialer.go the way we used to.

File: src/github.com/getlantern/checkfallbacks/verify.go (hunk 4 above)
Comment (fffw) [inline src/github.com/getlantern/checkfallbacks/verify.go]: For those listed url to test, 200 is the only expected code. `http.Client` will fetch the redirected page automatically.

File: src/github.com/getlantern/checkfallbacks/checkfallbacks.go (hunk 1 above)
Comment (fffw) [inline src/github.com/getlantern/checkfallbacks/checkfallbacks.go]: Yes that's really annoying. I guess the currently running `checkfallbacks` on cloudmaster is a previous version.

**Comment by fffw** [comment]
@uaalto Made a few updates as per your advice.

File: src/github.com/getlantern/checkfallbacks/checkfallbacks.go (hunk 1 above)
Comment (uaalto) [inline src/github.com/getlantern/checkfallbacks/checkfallbacks.go]: I saw it [here](https://github.com/getlantern/lantern/blob/valencia/src/github.com/getlantern/flashlight/client/chained.go#L98), but that's right, is application code. It's also [here](https://github.com/getlantern/lantern/blob/valencia/src/github.com/getlantern/flashlight/config/config.go#L578) and [here](https://github.com/getlantern/lantern/blob/valencia/src/github.com/getlantern/flashlight/client/reverseproxy.go#L33), so from the library design point of view, it is expected to be defined always in the user code.

**Comment by uaalto** [comment]
Ok, @fffw I finished with the review. Just that nil check I mentioned in the last comment and I'm ready to merge!

**Comment by fffw** [comment]
Fixed that nil checking, thanks @uaalto

**Comment by uaalto** [comment]
Thank you @fffw! This is going to be very useful for testing, indeed.
